src/test/java/de/retest/ui/descriptors/DefaultRetestIdProviderTest.java

```diff
@@ -0,0 +1,88 @@
+		final IdentifyingAttributes ident = createIdentAttributes(
+				"This is some very long sentence, that could be in a link text, or in some paragraph, and really is to long to be used as id." );
+		final String retestId = cut.getRetestId( ident );
+		assertThat( retestId.length() ).isLessThan( 20 );
+
+		// and cut should still be unique!
+		assertThat( retestId.length() ).isNotEqualTo( cut.getRetestId( ident ) );
+	}
+
+	@Test
+	public void too_long_words_should_be_cut() {
+		final DefaultRetestIdProvider cut = new DefaultRetestIdProvider();
+
+		// but also for single words
+		final IdentifyingAttributes ident = createIdentAttributes( "supercalifragilisticexpialidocious" );
+		final String retestId = cut.getRetestId( ident );
+		assertThat( retestId.length() ).isLessThan( 20 );
+
+		// and cut should still be unique!
+		assertThat( retestId.length() ).isNotEqualTo( cut.getRetestId( ident ) );
+	}
+
+	public IdentifyingAttributes createIdentAttributes( final String text ) {
+		final Collection<Attribute> attributes = createList( fromString( "/HTML/DIV" ), "DIV" );
+		attributes.add( new StringAttribute( "text", text ) );
+		final IdentifyingAttributes ident = new IdentifyingAttributes( attributes );
+		return ident;
+	}
+
+	@Test
+	public void should_always_be_unique() {
+		final DefaultRetestIdProvider cut = new DefaultRetestIdProvider();
+		final IdentifyingAttributes ident = createIdentAttributes( "a" );
+		final String retestId = cut.getRetestId( ident );
+		assertThat( retestId ).isNotEqualTo( cut.getRetestId( ident ) );
+		assertThat( retestId ).isNotEqualTo( cut.getRetestId( ident ) );
+		assertThat( retestId ).isNotEqualTo( cut.getRetestId( ident ) );
+		assertThat( retestId ).isNotEqualTo( cut.getRetestId( ident ) );
+	}
+
+	@Test
+	public void works_even_only_for_path_and_type() {
+		final DefaultRetestIdProvider cut = new DefaultRetestIdProvider();
+		final IdentifyingAttributes ident = create( fromString( "/HTML/DIV" ), "DIV" );
+		final String id1 = cut.getRetestId( ident );
+		final String id2 = cut.getRetestId( ident );
+		assertThat( id1 ).isNotEqualTo( id2 );
+	}
+
+	@Test
+	public void no_text_should_give_type() {
+		final DefaultRetestIdProvider cut = new DefaultRetestIdProvider();
+		final Collection<Attribute> attributes = createList( fromString( "/HTML/DIV" ), "DIV" );
+		attributes.add( new StringAttribute( "type", "DIV" ) );
+		attributes.add( new SuffixAttribute( "3" ) );
+		assertThat( cut.getRetestId( new IdentifyingAttributes( attributes ) ) ).isEqualTo( "div" );
+	}
+
+	@Test( expected = NullPointerException.class )
+	public void null_should_give_exception() {
+		final DefaultRetestIdProvider cut = new DefaultRetestIdProvider();
+		cut.getRetestId( null );
+	}
+
+	@Test( expected = NullPointerException.class )
+	public void null_path_should_give_exception() {
+		final DefaultRetestIdProvider cut = new DefaultRetestIdProvider();
+		cut.getRetestId( create( null, "DIV" ) );
```

Comment:
This is not behavior of the CUT, the `NullPointerException` is thrown by `PathAttribute`. The CUT has no check for a `null` path.

Reply:
True. But this behavior should be checked here non the less.